tests/pipelines/test_contrastive.py

```diff
@@ -0,0 +1,153 @@
+        )
+        valid_loader = torch.utils.data.DataLoader(contrastive_valid, batch_size=BATCH_SIZE)
+
+        # 2. model and optimizer
+        encoder = nn.Sequential(nn.Flatten(), nn.Linear(28 * 28, 16), nn.LeakyReLU(inplace=True))
+        projection_head = nn.Sequential(
+            nn.Linear(16, 16, bias=False), nn.ReLU(inplace=True), nn.Linear(16, 16, bias=True),
+        )
+
+        class ContrastiveModel(torch.nn.Module):
+            def __init__(self, model, encoder):
+                super(ContrastiveModel, self).__init__()
+                self.model = model
+                self.encoder = encoder
+
+            def forward(self, x):
+                emb = self.encoder(x)
+                projection = self.model(emb)
+                return emb, projection
+
+        model = ContrastiveModel(model=projection_head, encoder=encoder)
+
+        optimizer = Adam(model.parameters(), lr=LR)
+
+        # 3. criterion with triplets sampling
+        criterion = NTXentLoss(tau=0.1)
+
+        callbacks = [
+            dl.ControlFlowCallback(
+                dl.CriterionCallback(
+                    input_key="projection_left", target_key="projection_right", metric_key="loss"
+                ),
+                loaders="train",
+            ),
+            dl.SklearnModelCallback(
+                feature_key="embedding_left",
+                target_key="target",
+                train_loader="train",
+                valid_loaders="valid",
+                model_fn=RandomForestClassifier,
+                predict_method="predict_proba",
+                predict_key="sklearn_predict",
+                random_state=RANDOM_STATE,
+                n_estimators=50,
+            ),
+            dl.ControlFlowCallback(
+                dl.AccuracyCallback(
+                    target_key="target", input_key="sklearn_predict", topk_args=(1, 3)
+                ),
+                loaders="valid",
+            ),
+        ]
+
+        runner = dl.SelfSupervisedRunner()
+
+        logdir = "./logdir"
+        runner.train(
+            model=model,
+            engine=engine or dl.DeviceEngine(device),
+            criterion=criterion,
+            optimizer=optimizer,
+            callbacks=callbacks,
+            loaders={"train": train_loader, "valid": valid_loader},
+            verbose=False,
+            logdir=logdir,
+            valid_loader="train",
+            valid_metric="loss",
+            minimize_valid_metric=True,
+            num_epochs=TRAIN_EPOCH,
+        )
+
+        valid_path = Path(logdir) / "logs/valid.csv"
+        best_accuracy = max(
+            float(row["accuracy"]) for row in read_csv(valid_path) if row["accuracy"] != "accuracy"
+        )
+
+        assert best_accuracy > 0.6
+
+
+@mark.skipif(
+    not SETTINGS.ml_required or not SETTINGS.cv_required,
+    reason="catalyst[ml] and catalyst[cv] required",
+)
+def test_on_cpu():
+    train_experiment("cpu")
```

Comment:
could we add GPU-based tests also please?

Reply:
[Done](https://github.com/gr33n-made/catalyst/blob/simCLR/tests/pipelines/test_contrastive.py#L151-L245)